### PR TITLE
Support AD in cluster connection dialog

### PR DESCRIPTION
**Pull request by kevcunnane**
This is a preliminary PR for adding AD support in the controller. Key changes:
- Use token API to get auth token before using other endpoints. Note this needs server updates before it'll work as expected (will only checkin after verifying this)
- Add auth option in controller UI and plumb through connection save, load, and uses of the controller API
- Because the swagger spec is split in 2 created new file for the 2nd swagger spec including token auth endpoints. These come from a running cluster and instructions were updated to reflect this.

**New UI**
Changes: 
- Added authentication type field with "Basic" and "Windows Authentication" as the options
- Moved error notifications to the dialog instead of separate notification window. That's the recommended pattern
- Username / password aren't required for Windows Authentication. I couldn't find a way to change required status in form container on switching from Windows => Basic Auth so have error show on clicking OK. If there's a better way let me know!

image

### CODE REVIEW

File: extensions/big-data-cluster/src/bigDataCluster/controller/clusterControllerApi.ts [hunk 1]
```diff
@@ -23,65 +25,130 @@ class AuthConfiguration implements Authentication {
...
-			response: result.response as IHttpResponse,
-			endPoints: result.body as EndpointModel[]
-		};
-	} catch (error) {
-		throw new ControllerError(error, localize('bdc.error.getEndPoints', "Error retrieving endpoints from {0}", url));
+		let tokenApi = this.createTokenApi(ignoreSslVerification);
+		try {
+
+			// AD auth is available, login to keberos and
+			let host = getHostAndPortFromEndpoint(this._url).host;
```
Comment (Charles-Gagnon) [inline extensions/big-data-cluster/src/bigDataCluster/controller/clusterControllerApi.ts]: trailing comment

File: extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts [hunk 2]
```diff
@@ -7,40 +7,74 @@
...
+	public async onComplete(url: string, auth: AuthType, username: string, password: string, rememberPassword: boolean): Promise<void> {
 		try {
+
+			if (auth === 'basic') {
+				// Verify username and password as we can't make them required in the UI
+				if (!username) {
+					throw new Error(localize('err.controller.username.required', "Username is required"));
+				} else if (!password) {
+					throw new Error(localize('err.controller.password.required', "Password is required"));
+				}
```
Comment (Charles-Gagnon) [inline extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts]: What about just disabling the OK button until those fields are filled? Not sure if we can hook into the text updated events for the text box components though currently.

Also this means we won't have the red * anymore right? That's unfortunate if so, did you look into keeping that functionality?

File: extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts [hunk 3]
```diff
@@ -229,3 +229,7 @@ export class ControllerNode extends ControllerTreeNode {
 	}
 }
 
+function isControllerMatch(node: ControllerNode, url: string, auth: string, username: string): unknown {
```
Comment (Charles-Gagnon) [inline extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts]: >unknown [](start = 95, length = 7)

Shouldn't this be returning a boolean?

File: extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts [hunk 4]
```diff
@@ -142,17 +187,35 @@ export class AddControllerDialog {
...
 
+	private get authValue(): AuthType {
+		return (<azdata.CategoryValue>this.authDropdown.value).name as AuthType;
+	}
+
+	private onAuthChanged(): void {
+		let isBasic = this.authValue === 'basic';
+		this.usernameInputBox.enabled = isBasic;
+		this.passwordInputBox.enabled = isBasic;
+		if (!isBasic) {
```
Comment (Charles-Gagnon) [inline extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts]: Should probably disable the remember password checkbox too

File: extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts [hunk 5]
```diff
@@ -92,6 +127,12 @@ export class AddControllerDialog {
 					placeHolder: localize('textUrlLower', 'url'),
 					value: this.model.prefilledUrl
 				}).component();
+			this.authDropdown = this.uiModelBuilder.dropDown().withProperties({
+				values: this.model.authCategories,
+				value: this.model.prefilledAuth,
+				editable: false,
```
Comment (Charles-Gagnon) [inline extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts]: Make sure you correctly disable the fields if prefilledAuth isn't basic

File: extensions/big-data-cluster/src/bigDataCluster/auth.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as kerberos from 'kerberos';
+import * as vscode from 'vscode';
+
+export async function authenticateKerberos(hostname: string): Promise<string> {
+	const service = 'HTTP' + (process.platform === 'win32' ? '/' : '@') + hostname;
```
Comment (alanrenmsft) [inline extensions/big-data-cluster/src/bigDataCluster/auth.ts]: No HTTPS support?

Comment (kevcunnane) [reply extensions/big-data-cluster/src/bigDataCluster/auth.ts]: This is just how authentication protocol looks, it actually does use HTTPS as far as I know. However if I try to change here to "HTTPS" I get the following back:
```
Error during authentication - Miscellaneous failure (see text): Server (HTTPS/mssql.azdata.local@AZDATA.LOCAL) unknown
```

File: extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts (hunk 2 above)
Comment (kevcunnane) [reply extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts]: It is unfortunate. As mentioned I'm not aware of a way to do this without bigger changes. For now, I'm going to go with current implementation as I think it's fairly understandable.

File: extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts (hunk 4 above)
Comment (kevcunnane) [reply extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts]: Done

**Pull request merged**